dsl/read.go

```diff
@@ -83,64 +98,89 @@ func ReadHopsFileContents(hopsFileContent []FileContent) (*hcl.BodyContent, stri
+func getHopsDirFilePaths(root string) ([]string, error) {
+	seenPath := make(map[string]bool) // map of directories with a hops file
 
-	// read in the hops files and prepare for parsing
-	if info.IsDir() {
-		return readHopsDir(hopsPath)
-	}
-
-	content, err := os.ReadFile(hopsPath)
-	if err != nil {
-		return nil, err
-	}
+	var filePaths []string // list of file paths to be returned at the end (hops and other)
 
-	files := []FileContent{{
-		File:    hopsPath,
-		Content: content,
-	}}
-
-	return files, nil
-}
-
-// readHopsDir retrieves the content of all .hops files in a directory,
-// including sub directories, and returns then as a slice of fileContents
-func readHopsDir(dirPath string) ([]FileContent, error) {
-	filePaths := []string{}
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
 
-	err := filepath.WalkDir(dirPath, func(path string, d fs.DirEntry, err error) error {
+		// Get relative path from the root
+		relativePath, err := filepath.Rel(root, path)
 		if err != nil {
 			return err
 		}
-		// Exclude directories whose name starts with '..'
-		// This is because kubernetes configMaps create a set of symlinked
-		// directories for the mapped files and we don't want to pick those
-		// up. Those directories are named '..<various names>'
-		// Example:
-		// /my-config-map-dir
-		// |-- my-key -> ..data/my-key
-		// |-- ..data -> ..2023_10_19_12_34_56.789012345
-		// |-- ..2023_10_19_12_34_56.789012345
-		// |   |-- my-key
-		if d.IsDir() && strings.HasPrefix(d.Name(), "..") {
-			return filepath.SkipDir
+
+		if d.IsDir() {
+			// Exclude directories whose name starts with '..'
+			// This is because kubernetes configMaps create a set of symlinked
+			// directories for the mapped files and we don't want to pick those
+			// up. Those directories are named '..<various names>'
+			// Example:
+			// /my-config-map-dir
+			// |-- my-key -> ..data/my-key
+			// |-- ..data -> ..2023_10_19_12_34_56.789012345
+			// |-- ..2023_10_19_12_34_56.789012345
+			// |   |-- my-key
+			if strings.HasPrefix(d.Name(), "..") {
+				return filepath.SkipDir
+			}
+
+			// Skip any second children of the root (i.e. root/sub, yes, root/sub/sub, no)
+			if strings.Count(relativePath, string(filepath.Separator)) > 1 {
+				return filepath.SkipDir
+			}
+
+			return nil
 		}
-		if !d.IsDir() && filepath.Ext(path) == ".hops" {
-			filePaths = append(filePaths, path)
+
+		// Files in root (i.e root/a.hops), and anything other than first
+		// child directory of the root (i.e. root/sub/sub/a.hops) are skipped
+		if strings.Count(relativePath, string(filepath.Separator)) != 1 {
+			return nil
 		}
+
+		// Ensure only 1 .hops file per directory
+		if filepath.Ext(path) == HopsExt {
+			dirOnly := filepath.Dir(relativePath)
+			if seenPath[dirOnly] {
+				return fmt.Errorf("Only one hops file per directory allowed: %s", relativePath)
+			}
+
+			seenPath[dirOnly] = true
+		}
```

Comment:
Unsure on this one. In my head I wasn't going to enforce that at this level, so the technical capability still exists within hops to have several, but we simplify that in the UX and standard flow.

I don't know if we get much value from flat out blocking it?

Reply:
(Happy to discuss)

Reply:
Removed